src/contracts/BPool.sol

```diff
@@ -36,15 +34,29 @@ contract BPool is BToken, BMath, IBPool {
 
   /// @dev Prevents reentrancy in non-view functions
   modifier _lock_() {
-    require(!_mutex, 'ERR_REENTRY');
-    _mutex = true;
+    if (_getLock() != _MUTEX_FREE) {
+      revert BPool_Reentrancy();
+    }
+    _setLock(_MUTEX_TAKEN);
     _;
-    _mutex = false;
+    _setLock(_MUTEX_FREE);
   }
 
   /// @dev Prevents reentrancy in view functions
```

Comment:
perhaps could change description to "Throws an error if the reentrancy mutex is locked (doesn't modify the mutex)"